### PR TITLE
added new distributions to sampling operators. added operators for sa…

### PR DESCRIPTION
**Pull request by asmushetzel**
…mpling multiple distributions that are parametrized by input arguments(tensors)

This adds several other relevant distributions to mshadow sampler. These are the most important ones in a wide range of applications (Exponential/Poisson/Negative Binomial). Sampling with these distributions is only possible on the CPU though.
The existing uniform/gaussian sampling methods will still run on GPU with the prior existing code. 
This is just the part related to mshadow of a bigger pull request for the same branch name  in MxNet. Both pull requests actually belong together.  Please have a look at the other pull request from me.

### CODE REVIEW

**Comment by asmushetzel** [comment]
PR 5726 for dmlc/mxnet is the main part of the work and depending on this PR
(accidentally closed, will reopen)

File: mshadow/random.h [hunk 1]
```diff
@@ -68,47 +69,166 @@ class Random<cpu, DType> {
...
+  /*!
+   * \brief generate data from uniform [a,b)
+   * \param dst destination
+   * \param a lower bound of uniform
+   * \param b upper bound of uniform
+   * \tparam dim dimension of tensor
+   */
+  template<int dim, typename PType>
+  inline void SampleUniform(Tensor<cpu, dim, DType> *dst,
+                            PType a , PType b ) {
```
Comment (piiswrong) [inline mshadow/random.h]: keep the default values

File: mshadow/random.h [hunk 2]
```diff
@@ -68,47 +69,166 @@ class Random<cpu, DType> {
...
    * \brief generate data from standard gaussian
    * \param dst destination
    * \param mu mean variable
    * \param sigma standard deviation
    * \tparam dim dimension of tensor
    */
-  template<int dim>
+  template<int dim, typename PType>
   inline void SampleGaussian(Tensor<cpu, dim, DType> *dst,
-                             DType mu = 0.0f, DType sigma = 1.0f) {
```
Comment (piiswrong) [inline mshadow/random.h:99]: same

**Comment by piiswrong** [comment]
Please keep backward compatibility. Don't change interface or naming convention

File: mshadow/random.h (hunk 1 above)
Comment (asmushetzel) [reply mshadow/random.h]: Corrected this and retained default values for uniform/gaussian. Pushed it again. Will wait to deal with your remarks concerning the bigger mxnet part after this is pulled in as only then I can get a clean compile & test.

File: mshadow/random.h [hunk 3]
```diff
@@ -152,24 +272,39 @@ class Random<cpu, DType> {
   std::mt19937 rnd_engine_;
   /*! \brief random number seed used in random engine */
   unsigned rseed_;
-  // implementing generators.
-  inline void GenUniform(DType *dptr, index_t size, DType a, DType b) {
```
Comment (piiswrong) [inline mshadow/random.h:156]: why are these removed?

Comment (asmushetzel) [reply mshadow/random.h:156]: No longer referenced. All distributions use the unified generic generator SampleDistribution.

Comment (asmushetzel) [reply mshadow/random.h:156]: Should say: All distributions sampled on the CPU use SampleDistribution. GPU functionality is fully unchanged and still restricted to uniform/gaussian.
